### PR TITLE
Visualizer impelentation is provided

### PR DESCRIPTION
**Pull request by andrewt0301**
1. Implementation of the Visualizer.
2. Minimal tests (~80 % coverage), wich cover DOT code generation.

### CODE REVIEW

File: src/main/java/org/cqfn/patternika/visualizer/ImageVisualizer.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+ * @since 2021/02/19
+ */
+public class ImageVisualizer implements Visualizer {
+    /** Path to the Graphviz that performs rendering. */
+    private final String graphvizPath;
+    /** Name of the file to save the image.*/
+    private final String imagePath;
+    /** The action tree to be visualized. */
+    private final ActionTree tree;
+    /** Markers to highlight nodes. */
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/ImageVisualizer.java:28]: what kind of markers could be?

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package org.cqfn.patternika.visualizer.dot;
+
+import org.cqfn.patternika.ast.ActionType;
+
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * Action for a Dot graph.
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java:9]: DOT and Dot. Better to refer same names.
The file name for the class and the class name itself, I suggest to name DOT.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    @Override
+    public void accept(final StringBuilder builder) {
+        builder.append("  action_").append(index)
+               .append(" [shape=note color=")
+               .append(getColor())
+               .append(" label=<")
+               .append(type)
+               .append(">];\n");
+    }
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java:43]: Will we see the action node to which it relates to? It's very useful while debugging large patterns.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotMarkerStyle.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    public void accept(final StringBuilder builder) {
+        final boolean isSingle = markers.size() == 1;
+        if (isSingle) {
+            builder.append("style=\"rounded,filled\"");
+        } else {
+            builder.append("style=striped penwidth=2");
+        }
+        builder.append(" fillcolor=\"");
+        boolean isFirst = true;
+        for (final Integer marker : markers) {
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotMarkerStyle.java:65]: Please comment this logic (reasons)

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotMultiLink.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    private final int sourceIndex;
+    /** The target node name. */
+    private final String target;
+    /** The target node indexes. */
+    private final List<Integer> targetIndexes;
+
+    /**
+     * Constructor.
+     *
+     * @param source the source node name.
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotMultiLink.java:25]: What does "source code name" mean? Please, change the comment.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotNodeShape.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        this.nodeType = Objects.requireNonNull(nodeType);
+    }
+
+    /**
+     * Writes the shape to a string builder.
+     *
+     * @param builder the string builder.
+     */
+    @Override
+    public void accept(final StringBuilder builder) {
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotNodeShape.java:30]: The function name "accept" for the function that is about to write the shape to a string builder is not appropriate, don't you think?

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotNodeShape.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+                   .append(' ');
+        }
+    }
+
+    /**
+     * Returns the shape for the node type.
+     *
+     * @return the shape.
+     */
+    private String getShape() {
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotNodeShape.java:44]: What about other shapes? Actions, Identifiers, etc. If the function is only about traits, then it's necessary to change the function name.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.function.Consumer;
+
+/**
+ * Renders an abstract syntax tree to a Dot text.
+ *
+ * @since 2021/02/08
+ */
+public class DotVisualizer implements Visualizer {
+    /** Builds text for a Graphviz file. */
+    @SuppressWarnings("PMD.AvoidStringBufferField")
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java:26]: Please add a reason description to the rule check exception.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        }
+        for (final Action action : tree.getActionsByParent(node)) {
+            actionIndexes.put(action, ++lastIndex);
+            buildIndexes(action.getRef());
+            buildIndexes(action.getAccept());
+        }
+    }
+
+    /**
+     * Appends a node with all its actions and children.
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java:118]: childrenS?

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+            append(DotLink.newNodeToNode(parentIndex, currentIndex, childIndex));
+        }
+        if (node != null) {
+            appendChildren(node);
+            appendActions(node);
+        }
+    }
+
+    /**
+     * Returns the writer for the node style.
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java:144]: The comment for the function is unclear.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java (hunk 3 above)
Comment (andrewt0301) [reply src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java:43]: Yes.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java (hunk 2 above)
Comment (andrewt0301) [reply src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java:9]: DOT is the official name of the language. I agree.as a class name is a violation of cove conventions.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java (hunk 9 above)
Comment (andrewt0301) [reply src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java:118]: children - the valid plural form! "childrens" - no such word in English.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java (hunk 3 above)
Comment (andrewt0301) [reply src/main/java/org/cqfn/patternika/visualizer/dot/DotAction.java:43]: DotAction just write text. It is isolated from semantics.

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotNodeShape.java (hunk 7 above)
Comment (andrewt0301) [reply src/main/java/org/cqfn/patternika/visualizer/dot/DotNodeShape.java:44]: No other shapes. Default shape is used. If other custom shapes are needed, the method will be extended.

**Review by kniazkov — DISMISSED** [review]

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotNodeShape.java (hunk 7 above)
Comment (tiohlognm) [reply src/main/java/org/cqfn/patternika/visualizer/dot/DotNodeShape.java:43]: ok

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java (hunk 9 above)
Comment (tiohlognm) [reply src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java:117]: sorry I beg your pardon. thank you :)

File: src/main/java/org/cqfn/patternika/ast/Hole.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package org.cqfn.patternika.ast;
+
+/**
+ * This interface is to be supported by all holes (wildcard nodes).
```
Comment (tiohlognm) [inline src/main/java/org/cqfn/patternika/ast/Hole.java:4]: by all wildcard nodes? then can be simply "by all wildcard nodes". But, it is unclear for me which nodes are wildcard. Do we have a definition somewhere?

File: src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java (hunk 8 above)
Comment (tiohlognm) [reply src/main/java/org/cqfn/patternika/visualizer/dot/DotVisualizer.java:25]: the generateD?

File: src/main/java/org/cqfn/patternika/ast/Hole.java (hunk 11 above)
Comment (andrewt0301) [reply src/main/java/org/cqfn/patternika/ast/Hole.java:4]: "A pattern is a subtree of an original AST with the following additional features.
First, some of its nodes are replaced with \emph{holes}, which are wildcard nodes matching any nodes of the same type." - from my TR.

Comment (tiohlognm) [reply src/main/java/org/cqfn/patternika/ast/Hole.java:4]: please add this to the source code

Comment (andrewt0301) [reply src/main/java/org/cqfn/patternika/ast/Hole.java]: done

**Review by tiohlognm — APPROVED** [review]

**Review by silverCase — COMMENTED** [review]
All clear.

**Review by silverCase — APPROVED** [review]
All clear.

**Pull request merged**